subsys/bluetooth/host/att.c

```diff
@@ -2313,8 +2313,18 @@ static uint8_t att_error_rsp(struct bt_att_chan *chan, struct net_buf *buf)
 	BT_DBG("request 0x%02x handle 0x%04x error 0x%02x", rsp->request,
 	       sys_le16_to_cpu(rsp->handle), rsp->error);
 
-	/* Don't retry if there is no req pending or it has been cancelled */
-	if (!chan->req || chan->req == &cancel) {
+	/* Don't retry if there is no req pending or it has been cancelled.
+	 *
+	 * BLUETOOTH SPECIFICATION Version 5.2 [Vol 3, Part F]
+	 * page 1423:
+	 *
+	 * If an error code is received in the ATT_ERROR_RSP PDU that is not
+	 * understood by the client, for example an error code that was reserved
+	 * for future use that is now being used in a future version of the
+	 * specification, then the ATT_ERROR_RSP PDU shall still be considered to
+	 * state that the given request cannot be performed for an unknown reason.
+	 */
+	if (!chan->req || chan->req == &cancel || !rsp->error) {
 		err = BT_ATT_ERR_UNLIKELY;
```

Comment:
in the `!rsp->error` case, could we return one of the `BT_ATT_ERR_INVALID_PDU`, `BT_ATT_ERR_VALUE_NOT_ALLOWED` or `BT_ATT_ERR_OUT_OF_RANGE` codes? it would be more informative to the application IMO.
What do you think @jhedberg , is that allowed, or should we not mess with the error code returned from the peer (even if it's invalid in this particular case) ? I guess this comes back to our previous discussion of separating app err codes from protocol err codes.

Reply:
another point for @LingaoM , this only checks for err being `!=0`, but we could have an inline `is_valid_att_err()` function that actually checks the range we support (1 - 0x13).

Reply:
Yes, this is the same issue wrt local vs remote errors. Here we have a `uint8_t` instead of `int` so can't really distinguish. I'd just keep using the convention of `ERR_UNLIKELY` until we agree on a proper solution that's then applied everywhere in the code.

Regarding an `att_err_is_valid()` API, it should probably also include Application or Profile/Service errors, and only consider RFU values as invalid.
<img width="881" alt="Screenshot 2022-06-22 at 9 30 50" src="https://user-images.githubusercontent.com/1032570/174959062-fb328011-4e3a-4bd8-bbad-a1765cb060b5.png">